### ISSUE TITLE
Not able to authenticate user with google home

### ISSUE THREAD

**Issue by lsethi-xoriant** [doorkeeper-gem/doorkeeper#1107]
I am using doorkeeper as our oauth server , we are able to redirect user in sign in page when user invoke our  google home application after sign in not able to get token form oauth server.
We are getting an error **Bad response from IdP in Auth Code Exchange** 

But using https://developers.google.com/oauthplayground we are able to get token.

**Comment by nbulaj** [comment]
Hi @lsethi-xoriant . I don't understand how Doorkeeper is connected to Google Home. Maybe you could describe it in more details?

**Comment by lsethi-xoriant** [comment]
Hello @nbulaj. We have created a project in action on google (https://console.actions.google.com)
Over there is option call account linking which needs application id, secrets, Authorization URL,Token URL (Already created application on my oauth server and added required information for account linking in https://console.actions.google.com) .
Fallow steps as per link https://developers.google.com/actions/identity/oauth2-code-flow
When I try to test google action project in simulator **"I need to link your my test app account to Google. Is that okay?"**
Which is correct As I have done linking for my project. There is url generated into Debug option in simulator copy paste url into browser. This url redirect me to my oauth server where I have added email and pwd in sign in page . Ask for authorize the application after click on authorize I got an error **Bad response from IdP in Auth Code Exchange**

**Comment by lsethi-xoriant** [comment]
@nbulaj Issue got fixed as we have restrict our server access for few countries only. Google simulator request comes every time from different  geolocation .
BTW after getting an access_token and make request to the server I am facing 401 Error.
**Filter chain halted as :doorkeeper_authorize! rendered or redirected 2018 Completed 401 Unauthorized** 
Please take look of request as below
`2018-07-04T11:35:40.120272+00:00 app[web.1]: [ff95e1b2-5da3-4875-9ffd-c224a2cbcf4d]   Parameters: {"responseId"=>"fe0591a6-d34d-492b-8517-f904dd4a9b03", "queryResult"=>{"queryText"=>"GOOGLE_ASSISTANT_WELCOME", "action"=>"input.welcome", "parameters"=>{}, "allRequiredParamsPresent"=>true, "outputContexts"=>[{"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/google_assistant_welcome"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_screen_output"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_audio_output"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/google_assistant_input_type_keyboard"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_media_response_audio"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_web_browser"}], "intent"=>{"name"=>"projects/mi2-development-dba79/agent/intents/3fec194e-2b37-4dfd-968d-0e0613a3da39", "displayName"=>"Default Welcome Intent"}, "intentDetectionConfidence"=>1.0, "languageCode"=>"en-us"}, "originalDetectIntentRequest"=>{"source"=>"google", "version"=>"2", "payload"=>{"isInSandbox"=>true, "surface"=>{"capabilities"=>[{"name"=>"actions.capability.AUDIO_OUTPUT"}, {"name"=>"actions.capability.SCREEN_OUTPUT"}, {"name"=>"actions.capability.MEDIA_RESPONSE_AUDIO"}, {"name"=>"actions.capability.WEB_BROWSER"}]}, "requestType"=>"SIMULATOR", "inputs"=>[{"rawInputs"=>[{"query"=>"Talk to my test app", "inputType"=>"KEYBOARD"}], "intent"=>"actions.intent.MAIN"}], "user"=>{"lastSeen"=>"2018-07-04T11:34:24Z", "accessToken"=>"a99259b44d8bf1c97b7bea4c4fa61cb40b8ff9cc3df8a4e31b1346467da1e616", "locale"=>"en-US", "userId"=>"ABwppHFAnvgRPxVW3PXj1SVfRTI9-OZW-jibNmRbCdEfHyeCisVh_FLrprYa5_BlfcOpz0aDNGIcvzk8PtsC0WaNWmTt"}, "conversation"=>{"conversationId"=>"1530704136694", "type"=>"NEW"}}}, "session"=>"projects/mi2-development-dba79/agent/sessions/1530704136694", "handler"=>{"responseId"=>"fe0591a6-d34d-492b-8517-f904dd4a9b03", "queryResult"=>{"queryText"=>"GOOGLE_ASSISTANT_WELCOME", "action"=>"input.welcome", "parameters"=>{}, "allRequiredParamsPresent"=>true, "outputContexts"=>[{"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/google_assistant_welcome"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_screen_output"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_audio_output"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/google_assistant_input_type_keyboard"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_media_response_audio"}, {"name"=>"projects/mi2-development-dba79/agent/sessions/1530704136694/contexts/actions_capability_web_browser"}], "intent"=>{"name"=>"projects/mi2-development-dba79/agent/intents/3fec194e-2b37-4dfd-968d-0e0613a3da39", "displayName"=>"Default Welcome Intent"}, "intentDetectionConfidence"=>1.0, "languageCode"=>"en-us"}, "originalDetectIntentRequest"=>{"source"=>"google", "version"=>"2", "payload"=>{"isInSandbox"=>true, "surface"=>{"capabilities"=>[{"name"=>"actions.capability.AUDIO_OUTPUT"}, {"name"=>"actions.capability.SCREEN_OUTPUT"}, {"name"=>"actions.capability.MEDIA_RESPONSE_AUDIO"}, {"name"=>"actions.capability.WEB_BROWSER"}]}, "requestType"=>"SIMULATOR", "inputs"=>[{"rawInputs"=>[{"query"=>"Talk to my test app", "inputType"=>"KEYBOARD"}], "intent"=>"actions.intent.MAIN"}], "user"=>{"lastSeen"=>"2018-07-04T11:34:24Z", "accessToken"=>"a99259b44d8bf1c97b7bea4c4fa61cb40b8ff9cc3df8a4e31b1346467da1e616", "locale"=>"en-US", "userId"=>"ABwppHFAnvgRPxVW3PXj1SVfRTI9-OZW-jibNmRbCdEfHyeCisVh_FLrprYa5_BlfcOpz0aDNGIcvzk8PtsC0WaNWmTt"}, "conversation"=>{"conversationId"=>"1530704136694", "type"=>"NEW"}}}, "session"=>"projects/mi2-development-dba79/agent/sessions/1530704136694"}}
2018-07-04T11:35:40.122250+00:00 app[web.1]: [ff95e1b2-5da3-4875-9ffd-c224a2cbcf4d] Filter chain halted as :doorkeeper_authorize! rendered or redirected
2018-07-04T11:35:40.122483+00:00 app[web.1]: [ff95e1b2-5da3-4875-9ffd-c224a2cbcf4d] Completed 401 Unauthorized in 2ms (ActiveRecord: 0.0ms)
2018-07-04T11:35:40.123637+00:00 heroku[router]: at=info method=POST path="/api/google/handler" host=uslab-guestbook.herokuapp.com request_id=ff95e1b2-5da3-4875-9ffd-c224a2cbcf4d fwd="35.232.38.96" dyno=web.1 connect=1ms service=10ms status=401 bytes=434 protocol=https`

**Comment by rept** [comment]
@lsethi-xoriant did you solve your last issue?
I'm getting the exact same thing with Google assistant & doorkeeper: Filter chain halted as :doorkeeper_authorize! rendered or redirected Completed 401

**Comment by lsethi-xoriant** [comment]
@rept  Yes I fixed this issue using own authentication method. instead of doorkeeper_authorize!
